README: Add YAML config file to example; add info on file loaders

pauln opened the pull request:
As YAML is a widely-used format for config files, it's great to see that `uconfig` can support it simply by feeding in an existing YAML (un)marshaller.  It's not immediately obvious that this is possible, though; adding it to the README examples (and tests, because why not?) should help make it more obvious that it's supported out of the box.

omeid commented:
Thanks a lot for your contribution!

I agree that the docs could be improve a lot and a `yaml` example is definitely a good thing to have!

I am just not sure about adding it as dependency even if it is just for the tests, a fair bit of thought has been put into making uconfig light keep the dependencies minimal.

What do you think about adding the YAML example in the docs with a bit more explanation on how it works? Or perhaps even in the file plugin readme?

pauln commented:
I definitely get (and appreciate!) the desire to keep it lean - but from that perspective, you're already importing TOML purely for testing.  I guess you'll need to decide whether you're happy to bring in external dependencies just to test against (in which case this PR should be fine as it stands) or you'd rather keep it as lean as possible (in which case perhaps the TOML dependency / test should be removed).

Personally, I think that clear documentation which includes examples should be sufficient.  Including third-party libs just to test against them does seem excessive, so perhaps the TOML test (and thus dependency) should be stripped out, in favour of expanded / improved documentation and examples.  I'm happy to help flesh out the docs / examples, if this is the way you'd like to go.

I guess you could also set up a separate repo dedicated to (file plugin?) tests, which could include as many external dependencies as you like without bloating the main `uconfig` package; as long as that repo depends on the latest `uconfig`, it should still make for a reasonably clean and straightforward development process.

omeid commented:
The existence of toml as test dependency is a mistake. It should not be there, I believe the json encoder which is part of the library is enough to test the behaviour of the file plugin. You're right in that it should be removed 👍 

And I totally agree that good documents should be enough :)

If you would update your pull request to remove toml and extend the documentation, I would merge it right away!

pauln commented:
I've fixed and updated the example in the README.  Although I had intended to add multiple config files in different formats to the example, I ended up going with a single YAML file so that the example stays reasonably short.

I also added a new section to the README (above the example) which gives a bit of information about file loaders, including a (very brief) list of unmarshallers which are known to work; that should hopefully suffice - and the list can easily be added to as and when additional unmarshallers are tested or reported to work.

omeid commented:
Thank you!
